pyqir-generator/pyqir/generator/qis.pyi

```diff
@@ -0,0 +1,169 @@
+        """
+        Builds a Hadamard gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def m(self, qubit: Qubit, result: Ref) -> None:
+        """
+        Builds a measurement operation.
+
+        :param qubit: The qubit to measure.
+        :param result: A result reference where the measurement result will be
+                       written to.
+        """
+        ...
+
+    def reset(self, qubit: Qubit) -> None:
+        """
+        Builds a reset operation.
+
+        :param qubit: The qubit to reset.
+        """
+        ...
+
+    def rx(self, theta: float, qubit: Qubit) -> None:
+        """
+        Builds a rotation gate about the :math:`x` axis.
+
+        :param theta: The angle to rotate by.
+        :param qubit: The qubit to rotate.
+        """
+        ...
+
+    def ry(self, theta: float, qubit: Qubit) -> None:
+        """
+        Builds a rotation gate about the :math:`y` axis.
+
+        :param theta: The angle to rotate by.
+        :param qubit: The qubit to rotate.
+        """
+        ...
+
+    def rz(self, theta: float, qubit: Qubit) -> None:
+        """
+        Builds a rotation gate about the :math:`z` axis.
+
+        :param theta: The angle to rotate by.
+        :param qubit: The qubit to rotate.
+        """
+        ...
+
+    def s(self, qubit: Qubit) -> None:
+        """
+        Builds an :math:`S` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def s_adj(self, qubit: Qubit) -> None:
+        """
+        Builds an adjoint :math:`S` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def t(self, qubit: Qubit) -> None:
+        """
+        Builds a :math:`T` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def t_adj(self, qubit: Qubit) -> None:
+        """
+        Builds an adjoint :math:`T` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def x(self, qubit: Qubit) -> None:
+        """
+        Builds a Pauli :math:`X` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def y(self, qubit: Qubit) -> None:
+        """
+        Builds a Pauli :math:`Y` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def z(self, qubit: Qubit) -> None:
+        """
+        Builds a Pauli :math:`Z` gate.
+
+        :param qubit: The target qubit.
+        """
+        ...
+
+    def if_result(
+        self,
+        result: Ref,
+        one: Callable[[], None] = ...,
+        zero: Callable[[], None] = ...,
+    ) -> None:
+        """
+        Builds a conditional branch on the result of a measurement.
+
+        Dereferences the result reference, then evaluates the instructions
+        built by ``one`` if the result is one, or the instructions built
+        by ``zero`` if the result is zero. The one and zero callables should use
+        this builder to build instructions.
```

Comment:
Regarding the need to use the same builder within the callables, is there any way to enforce that? Just thinking out loud, could it be part of the convention that the builder passes itself into the callables? Nothing for this PR, just food for thought.

Reply:
I see in the Rust code that it's already setup to use the same generator. Is this comment still necessary? Is it possible from python to pass a different builder such that this expectation gets violated?

Reply:
> Regarding the need to use the same builder within the callables, is there any way to enforce that?

Not easily, I don't think. You could have any number of modules and builders in scope and accidentally build on the wrong one.

For example:

```python
mod1 = SimpleModule(...)
mod2 = SimpleModule(...)
qis1 = BasicQisBuilder(mod1.builder)
qis2 = BasicQisBuilder(mod2.builder)

# Oops, wrong builder in one branch.
qis1.if_result(mod1.results[0], lambda: qis2.x(mod2.qubits[0]))
```

> Just thinking out loud, could it be part of the convention that the builder passes itself into the callables?

Yes, you could have the convention that the branch generators are of the form `lambda qis: qis.x(mod.qubits[0])` or `lambda builder: BasicQisBuilder(builder).x(mod.qubits[0])` instead of a closure over the builder. However, this would only be a hint, not a requirement that the given builder should be used.